### PR TITLE
Add purgeChunk command that re-generates a single chunk

### PR DESCRIPTION
**Pull request by msteiger**

### CODE REVIEW

**Comment by Cervator** [comment]
Merged as it seems to do ... something? Without misbehaving :-)

I'm not sure exactly what the intent is, I figured it would delete and entirely recreate an existing chunk, but when I try it on 0 0 0 after placing a few things I get some unavailable chunk log spam and the screen seems to regen, but the stuff I placed remains there.

Same in headless and if I move one chunk off just for testing.

File: engine/src/main/java/org/terasology/logic/console/commands/ServerCommands.java [hunk 1]
```diff
@@ -151,6 +152,11 @@ public void save() {
         storageManager.requestSaving();
     }
 
+    @Command(shortDescription = "Deletes the specified chunk and re-generate it", runOnServer = true)
+    public void purgeChunk(@CommandParam("x") int x, @CommandParam("y") int y, @CommandParam("z") int z) {
+        chunkProvider.purgeChunk(new Vector3i(x, y, z));
```
Comment (Cervator) [inline engine/src/main/java/org/terasology/logic/console/commands/ServerCommands.java]: Suggestion for future commit: Make this print something to the console, maybe the boolean value returned to tell the player if the purge did something or not?

**Comment by flo** [comment]
If you start the game without "-noSaveGames". this command will unload the chunk and all its entities to memory (and eventually to disk) just to restore them instantly afterwards. So you just see a white flash.

If you use "-noSaveGames" then the command generates the chunk again or loads it from disk.

Maybe the command should print a hint that it is only usefull together with "-noSaveGames"?

**Comment by msteiger** [comment]
Interesting - I tried to reproduce the problem for a while now. With save games disabled, it works fine.
Maybe we should also delete the chunk from disk instead to have consistent behavior. What do you think?

**Comment by flo** [comment]
I think that would be a good idea. It is however more than a simple file delete:

Things to consider:
1. Per default chunks get stored in zips, a simple file delete is not possible. It is however also possible to store chunks as single files.
2. When you modify the files you should get the write lock for the directory which the chunks found in the ReadWriteStorageManager
3. There are 2 maps in the ReadWriteStorageManager that contain unloaded chunks before they end up in the disk
4. It would be slightly more correct, to destroy the entities instead of unloading them.

**Comment by msteiger** [comment]
Thanks for the info. I see two more options we could pursue:

* Don't write the chunk to the StorageManager - i.e. don't call `deactivateChunk()`. I guess this would make sense as we don't want to save changes we're rejecting.
* Don't read the chunk from the StorageManager - i.e. explicitly trigger a create op. instead of loading it. This would effectively ignore any previously stored state of the chunk on disk. Does it make sense to do that?

**Comment by flo** [comment]
About option 1 of your last comment: 
* Even if we don't call deactivateChunk, the chunk may still be stored in the 2 maps I mentioned if it got un- and reloaded since the last autosave through player movement. Also the copy on the disk is in average only 30s old. In addition not calling deactiveChunk would mean that some life cycle even'ts don't get sent.
About option 2:
* Destroying all entities in the chunk and doing that might be a option, although it is tricky as the chunk loading happens mostly off the main thread. So you would have to pass a flag to the task and somehow make sure that even if the task gets canceld that the new task will again perform the same reloading. Maybe we could create will a set in the storage manager with chunks that got pruned and need to be reloaded but it is also trick with concurrency.

Thinking about it I think we should do a correct purgeChunk implementation that also deletes the chunk from the disk. It is not so hard to do and is the most usefull and propably final implementation.
The deleting should ideally happen in a transaction like the saving so that it is crash save.

This save transaction concept works currently like this:
1. At first a folder called "unfinished-save-transaction" is created. In this folder copies of all modified files as well as new files will be placed.
   * If it crashes till that stage, all changes till then can simply be reverted by deleting the unfinished-save-transaction folder.
2. Then it renames that folder atomically to "unmerged-changes". 
 * If it crashes now step 3 will be done at the next start
3. It applies the changes from the unmerged-changes to the actual save game. e.g. if there is a file in unmerged-changes it will override the file with the same relative path in the save game folder.
* If a crash happens step 3 will be resumed.

Deleting a chunk from a zip is as simple as placing copying the zip to unfinished-save-transaction and removing it from the copy. Then step 2 and 3 are done as usually.

We could extend the merge concept by introducing the concept of having files that end with "-deletemarker" in the unmerged-changes folder be interpreted in stage 3 as a delete wish of a file with the same name except for the suffix "-deletemarker".

In the case of the purgeChunk command, I think it would be okay to do all 3 steps on the main thread but with the write lock. hmm but the lock does not get hold long enough to prevent a conflict with the save transaction which is preparing the folder with the files without lock. 

That would mean we have to do the chunk removal within the save transaction. As a result we would have to let the storage manager know that a chunk should be deleted at the next possible time. e.g. it would have to have at least 1 map in which it stores that information. Probably 2 like we currently have 2 maps for unloaded chunks.

Why we need 2 sets:
1. you request the purge of a chunk. its position goes into a set chunksToDelete
2. the saving starts, the position gets moved into a second set chunksBeingDeleted
3. the chunk might be pruged again, the possition gets added to chunksToDelete
4. the saving is done, the chunks position get removed from chunksBeingDeleted but stays in chunksToDelete.
For the decsion in point 4 two sets are helpfull, even if double deleting is not an issue.

The chunksToDelete set would then be passed from the ReadWriteSotrageManager to the created SaveTransaction which is then preparing the folder with the -deletemarker suffixes or zips with less chunks.

**Comment by msteiger** [comment]
Thanks for the details! The motivation behind this PR is a command that is used for debugging only - I don't see relevance for the actual game. 

As a consequence, I wouldn't go for the actual removal of chunks from zip files, because I don't see a use case for this scenario. Either use the the read-only manager or reload a ~ 30sec old version of the chunk. As long as it doesn't break things, I'm happy with it.

About option 1: I did that in #1530 - I believe that all life-cycle events are handled properly already. However, I added a method `deleteChunk` to delete all entities in the chunk and clean up the two maps to catch the corner cases. See 7ba4c83.

Does this cover all issues? I tested with both storage managers and couldn't provoke any incorrect behavior.

**Comment by flo** [comment]
I would like if we go either for the full version that also deletes a chunk (use case: Cleanup of Griefing, Manual regeneration of certain chunks after world generator changes, chunk deletion for save game size reduction, regeneration of "a mining area" on a server) or let the command explictly state that it does not work if saving is enabled.

**Comment by msteiger** [comment]
Sounds reasonable to me. I will tweak the PR over the weekend.

**Comment by msteiger** [comment]
I've updated the PR: the `deleteChunk` commit was reverted and the command fires only when writing save games is disabled.

**Pull request merged**